ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveSortLimitRemoveRule.java

```diff
@@ -41,8 +42,17 @@ public boolean matches(RelOptRuleCall call) {
     final HiveSortLimit sortLimit = call.rel(0);
 
     Double maxRowCount = call.getMetadataQuery().getMaxRowCount(sortLimit.getInput());
-    if (maxRowCount != null &&(maxRowCount <= 1)) {
-      return true;
+    if (maxRowCount != null) {
+      if (sortLimit.getFetchExpr() != null) {
+        // we have LIMIT
+        int limit = RexLiteral.intValue(sortLimit.getFetchExpr());
+        if (maxRowCount <= limit) {
```

Comment:
if I understand this correctly: this condition seems to be heavily building on that `MaxRowCount` is an overestimation of the actual rowcount - at any time that's not the case and it's an underestimation - we may end up removing the limit.

Because of the above: I think it would make sense to add a toggle to consider the removal in these cases or not...

Reply:
@kgyrtkirk `MaxRowCount` is guaranteed to return the number of rows which will be >= the actual number of rows, therefore we should never end up removing LIMIT when we shoudn't

Reply:
I agree; I've taken a closer look and for the cases I was considering the `maxRowCount` ends up being infinity (the default for all tables is infinity) - and no filter expr changes that.